File: crawler/src/protocol.ts
/**
 * Rust 与 crawler 之间的 NDJSON IPC 协议。
 *
 * 两个方向的每一条消息都是一个单行 JSON 对象。Rust 侧的同名类型定义位于
 * `src-tauri/src/sidecar.rs`，修改此处时必须同步修改。
 */

/** 文档第 36 条规定的任务状态。 */
export type TaskState =
  | "Pending"
  | "Fetching"
  | "Rendering"
  | "Extracting"
  | "Converting"
  | "Saving"
  | "Completed"
  | "Skipped"
  | "Failed"
  | "Blocked"
  | "RequiresLogin";

/** 文档第 37 条规定的错误类型。 */
export type CrawlErrorKind =
  | "InvalidURL"
  | "NetworkError"
  | "Timeout"
  | "HTTPError"
  | "AccessDenied"
  | "RateLimited"
  | "CaptchaDetected"
  | "ChallengeDetected"
  | "LoginRequired"
  | "PageRenderFailed"
  | "ContentExtractionFailed"
  | "MarkdownConversionFailed"
  | "PDFConversionFailed"
  | "SaveFailed"
  | "DuplicateURL";

/** 文档第 32 条规定的输出格式，默认 Markdown。 */
export type OutputFormat = "markdown" | "pdf" | "both";

/**
 * 图片保存策略（文档第 34 条，经作者确认为两者同时支持，由 UI 开关控制）。
 *
 * - `remote`：Markdown 中保留远程图片 URL
 * - `local`：图片下载到本地，Markdown 使用相对路径引用
 */
export type ImageStrategy = "remote" | "local";

export interface CrawlOptions {
  /** 输出格式。 */
  format: OutputFormat;
  /** 图片保存策略。 */
  imageStrategy: ImageStrategy;
  /**
   * 是否启用多页自动续页。文档第 22 条：默认关闭，上限 5 页，
   * 无法可靠判断下一页时应停止并提示，不得猜测。
   */
  followPagination: boolean;
  /** 自动续页上限。 */
  maxPagination: number;
  /**
   * 是否将同一内容集合的各页输出为独立文件。
   * 文档第 21 条：默认接续为单文件，仅在用户选择时独立输出。
   */
  separateOutput: boolean;
  /** 是否遵循 robots.txt。用户显式输入的 URL 始终豁免。 */
  obeyRobots: boolean;
}

export const DEFAULT_CRAWL_OPTIONS: CrawlOptions = {
  format: "markdown",
  imageStrategy: "remote",
  followPagination: false,
  maxPagination: 5,
  separateOutput: false,
  obeyRobots: true,
};

/** 一个待抓取条目。`raw` 用于实际请求，`key` 仅用于查重。 */
export interface CrawlTarget {
  raw: string;
  key: string;
  /**
   * 注册域（eTLD+1），由 Rust 侧用 PSL 计算后下发。
   *
   * 多页发现的「同站范围限制」以此为基准，避免在 Node 侧重复实现一遍
   * 公共后缀列表逻辑，也保证两端的站点判定口径完全一致。
   */
  siteKey: string;
}

/** 一条抓取结果。 */
export interface PageResult {
  /** 与请求中的 `key` 对应，供 Rust 侧关联任务。 */
  key: string;
  /** 实际抓取的 URL。 */
  url: string;
  title: string;
  /** 正文 Markdown，不含标题与元数据块。 */
  markdown: string;
  /** 完成抓取的时间戳，格式 `YYYY-MM-DD HH:mm:ss`。 */
  crawledAt: string;
  /** 该页面是否由 Playwright 渲染得到（用于向用户说明耗时来源）。 */
  rendered: boolean;
  /** 多页接续中，同属一个内容集合的页面按此序号排列。 */
  sequence: number;
}

/** crawler → Rust 的消息。 */
export type OutboundMessage =
  | { type: "ready" }
  | {
      type: "progress";
      key: string;
      url: string;
      state: TaskState;
      step: string;
      progress: number;
    }
  | {
      type: "result";
      key: string;
      url: string;
      title: string;
      markdown: string;
      crawledAt: string;
      rendered: boolean;
      sequence: number;
    }
  | {
      /** 对 `render-pdf` 的应答。 */
      type: "pdf";
      id: string;
      /** base64 编码的 PDF 字节。 */
      pdfBase64: string;
    }
  | {
      /** 对 `render-pdf` 的失败应答。 */
      type: "pdf-error";
      id: string;
      message: string;
      detail?: string;
    }
  | {
      type: "error";
      key: string;
      url: string;
      errorKind: CrawlErrorKind;
      message: string;
      detail?: string;
    }
  | { type: "skipped"; key: string; url: string; reason: string }
  | { type: "done"; succeeded: number; failed: number; skipped: number }
  | { type: "log"; level: "info" | "warn" | "error"; message: string }
  | { type: "login-opened"; domain: string }
  | { type: "login-saved"; domain: string }
  | { type: "login-closed"; domain: string; saved: boolean };

/** Rust → crawler 的消息。 */
export type InboundMessage =
  | {
      cmd: "crawl";
      targets: CrawlTarget[];
      options: CrawlOptions;
      /** 登录态目录，由 Rust 解析后下发。 */
      authDir: string;
      /**
       * 图片暂存目录。`local` 图片策略下，下载的图片先写入此处，
       * 待 Rust 决定最终文档名后再搬运到用户选择的保存位置。
       */
      stagingDir: string;
      /** Playwright 浏览器可执行文件路径。 */
      browserPath?: string;
    }
  | { cmd: "login"; domain: string; startUrl: string; authDir: string; browserPath?: string }
  | {
      /**
       * 将已组装好的 Markdown 文档渲染为 PDF。
       *
       * 文档组装（含多页接续与元数据块）由 Rust 侧完成，crawler 只负责
       * Markdown → HTML → PDF 的渲染，保证 PDF 与 Markdown 内容一致。
       */
      cmd: "render-pdf";
      id: string;
      markdown: string;
      title: string;
      browserPath?: string;
    }
  | { cmd: "ping" }
  | { cmd: "shutdown" };

/** 单行 JSON 编解码。 */
export function encode(msg: OutboundMessage): string {
  return JSON.stringify(msg) + "\n";
}

export function decode(line: string): InboundMessage | null {
  const trimmed = line.trim();
  if (!trimmed) return null;
  try {
    return JSON.parse(trimmed) as InboundMessage;
  } catch {
    return null;
  }
}

File: src-tauri/src/protocol.rs
use serde::{Deserialize, Serialize};

use crate::error::CrawlErrorKind;

/// 与 `crawler/src/protocol.ts` 一一对应的 IPC 消息定义。
///
/// 修改此处时必须同步修改 TypeScript 侧，反之亦然。

/// 文档第 36 条规定的任务状态。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum TaskState {
    Pending,
    Fetching,
    Rendering,
    Extracting,
    Converting,
    Saving,
    Completed,
    Skipped,
    Failed,
    Blocked,
    RequiresLogin,
}

impl TaskState {
    /// 面向用户的中文说明。
    pub fn label(self) -> &'static str {
        match self {
            Self::Pending => "等待中",
            Self::Fetching => "获取页面",
            Self::Rendering => "浏览器渲染",
            Self::Extracting => "提取正文",
            Self::Converting => "转换格式",
            Self::Saving => "保存文件",
            Self::Completed => "已完成",
            Self::Skipped => "已跳过",
            Self::Failed => "失败",
            Self::Blocked => "被阻止",
            Self::RequiresLogin => "需要登录",
        }
    }

    /// 该状态是否代表任务已结束。
    pub fn is_terminal(self) -> bool {
        matches!(
            self,
            Self::Completed | Self::Skipped | Self::Failed | Self::Blocked | Self::RequiresLogin
        )
    }
}

/// 文档第 32 条规定的输出格式。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum OutputFormat {
    Markdown,
    Pdf,
    Both,
}

impl OutputFormat {
    pub fn writes_markdown(self) -> bool {
        matches!(self, Self::Markdown | Self::Both)
    }

    pub fn writes_pdf(self) -> bool {
        matches!(self, Self::Pdf | Self::Both)
    }
}

/// 图片保存策略（文档第 34 条，经作者确认为两者同时支持）。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ImageStrategy {
    Remote,
    Local,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CrawlOptions {
    pub format: OutputFormat,
    pub image_strategy: ImageStrategy,
    pub follow_pagination: bool,
    pub max_pagination: u32,
    pub separate_output: bool,
    pub obey_robots: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CrawlTarget {
    pub raw: String,
    pub key: String,
    #[serde(rename = "siteKey")]
    pub site_key: String,
}

// ---------------------------------------------------------------------------
// Rust → crawler
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "cmd", rename_all = "kebab-case")]
pub enum Outbound {
    Crawl {
        targets: Vec<CrawlTarget>,
        options: CrawlOptions,
        #[serde(rename = "authDir")]
        auth_dir: String,
        #[serde(rename = "stagingDir")]
        staging_dir: String,
        #[serde(rename = "browserPath", skip_serializing_if = "Option::is_none")]
        browser_path: Option<String>,
    },
    Login {
        domain: String,
        #[serde(rename = "startUrl")]
        start_url: String,
        #[serde(rename = "authDir")]
        auth_dir: String,
        #[serde(rename = "browserPath", skip_serializing_if = "Option::is_none")]
        browser_path: Option<String>,
    },
    RenderPdf {
        id: String,
        markdown: String,
        title: String,
        #[serde(rename = "browserPath", skip_serializing_if = "Option::is_none")]
        browser_path: Option<String>,
    },
    Ping,
    Shutdown,
}

// ---------------------------------------------------------------------------
// crawler → Rust
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
pub enum Inbound {
    Ready,
    Progress {
        key: String,
        url: String,
        state: TaskState,
        step: String,
        progress: f64,
    },
    Result {
        key: String,
        url: String,
        title: String,
        markdown: String,
        #[serde(rename = "crawledAt")]
        crawled_at: String,
        rendered: bool,
        sequence: u32,
    },
    Error {
        key: String,
        url: String,
        #[serde(rename = "errorKind")]
        error_kind: CrawlErrorKind,
        message: String,
        #[serde(default)]
        detail: Option<String>,
    },
    Skipped {
        key: String,
        url: String,
        reason: String,
    },
    Done {
        succeeded: u32,
        failed: u32,
        skipped: u32,
    },
    Log {
        level: String,
        message: String,
    },
    #[serde(rename = "login-opened")]
    LoginOpened { domain: String },
    #[serde(rename = "login-saved")]
    LoginSaved { domain: String },
    #[serde(rename = "login-closed")]
    LoginClosed { domain: String, saved: bool },
    Pdf {
        id: String,
        #[serde(rename = "pdfBase64")]
        pdf_base64: String,
    },
    #[serde(rename = "pdf-error")]
    PdfError {
        id: String,
        message: String,
        #[serde(default)]
        detail: Option<String>,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn 十一种任务状态齐备且名称与文档一致() {
        let all = [
            TaskState::Pending,
            TaskState::Fetching,
            TaskState::Rendering,
            TaskState::Extracting,
            TaskState::Converting,
            TaskState::Saving,
            TaskState::Completed,
            TaskState::Skipped,
            TaskState::Failed,
            TaskState::Blocked,
            TaskState::RequiresLogin,
        ];
        assert_eq!(all.len(), 11);
        for state in all {
            let json = serde_json::to_string(&state).unwrap();
            assert_eq!(json, format!("\"{:?}\"", state));
        }
    }

    #[test]
    fn 终结状态判定正确() {
        assert!(TaskState::Completed.is_terminal());
        assert!(TaskState::Failed.is_terminal());
        assert!(TaskState::Blocked.is_terminal());
        assert!(!TaskState::Fetching.is_terminal());
        assert!(!TaskState::Pending.is_terminal());
    }

    #[test]
    fn 输出格式的写出判定() {
        assert!(OutputFormat::Markdown.writes_markdown());
        assert!(!OutputFormat::Markdown.writes_pdf());
        assert!(OutputFormat::Pdf.writes_pdf());
        assert!(!OutputFormat::Pdf.writes_markdown());
        assert!(OutputFormat::Both.writes_markdown());
        assert!(OutputFormat::Both.writes_pdf());
    }

    #[test]
    fn 出站_crawl_消息序列化为_camelCase() {
        let msg = Outbound::Crawl {
            targets: vec![CrawlTarget {
                raw: "https://example.com/a".into(),
                key: "https://example.com/a".into(),
                site_key: "example.com".into(),
            }],
            options: CrawlOptions {
                format: OutputFormat::Markdown,
                image_strategy: ImageStrategy::Remote,
                follow_pagination: false,
                max_pagination: 5,
                separate_output: false,
                obey_robots: true,
            },
            auth_dir: "C:/auth".into(),
            staging_dir: "C:/staging".into(),
            browser_path: None,
        };

        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains("\"cmd\":\"crawl\""));
        assert!(json.contains("\"authDir\""));
        assert!(json.contains("\"stagingDir\""));
        assert!(json.contains("\"siteKey\""));
        assert!(json.contains("\"imageStrategy\":\"remote\""));
        assert!(json.contains("\"followPagination\""));
        // browserPath 为空时不应出现
        assert!(!json.contains("browserPath"));
    }

    #[test]
    fn 入站_progress_消息可解析() {
        let line = r#"{"type":"progress","key":"k","url":"https://e.com/a","state":"Fetching","step":"正在获取页面","progress":0.15}"#;
        let msg: Inbound = serde_json::from_str(line).unwrap();
        match msg {
            Inbound::Progress { state, progress, .. } => {
                assert_eq!(state, TaskState::Fetching);
                assert!((progress - 0.15).abs() < f64::EPSILON);
            }
            other => panic!("期望 Progress，实际为 {other:?}"),
        }
    }

    #[test]
    fn 入站_错误消息可解析且错误类型与文档一致() {
        let line = r#"{"type":"error","key":"k","url":"https://e.com/a","errorKind":"CaptchaDetected","message":"检测到验证码","detail":"页面包含 reCAPTCHA 资源"}"#;
        let msg: Inbound = serde_json::from_str(line).unwrap();
        match msg {
            Inbound::Error { error_kind, detail, .. } => {
                assert_eq!(error_kind, CrawlErrorKind::CaptchaDetected);
                assert!(error_kind.is_defense_mechanism());
                assert!(detail.is_some());
            }
            other => panic!("期望 Error，实际为 {other:?}"),
        }
    }

    #[test]
    fn 入站登录相关消息使用连字符命名() {
        let opened: Inbound =
            serde_json::from_str(r#"{"type":"login-opened","domain":"example.com"}"#).unwrap();
        assert!(matches!(opened, Inbound::LoginOpened { .. }));

        let closed: Inbound = serde_json::from_str(
            r#"{"type":"login-closed","domain":"example.com","saved":true}"#,
        )
        .unwrap();
        assert!(matches!(closed, Inbound::LoginClosed { saved: true, .. }));
    }

    #[test]
    fn 出站控制消息序列化正确() {
        assert_eq!(serde_json::to_string(&Outbound::Ping).unwrap(), r#"{"cmd":"ping"}"#);
        assert_eq!(
            serde_json::to_string(&Outbound::Shutdown).unwrap(),
            r#"{"cmd":"shutdown"}"#
        );
    }
}
